Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift

```diff
@@ -14,17 +14,170 @@ import SKSupport
+    errorHandler: @escaping (Error) -> Void,
+    handlerQueue: DispatchQueue)
+  {
+    self.messageHandler = messageHandler
+    self.errorHandler = errorHandler
+    self.handlerQueue = handlerQueue
+    self.jsonDecoder = JSONDecoder()
+    jsonDecoder.userInfo[.messageRegistryKey] = messageRegistry
+  }
+}
+
+extension JSONRPCConnectionBridge: ChannelDuplexHandler {
+
+  typealias InboundIn = ByteBuffer
+  typealias InbountOut = JSONRPCMessage
+  typealias OutboundIn = (JSONRPCMessage, OutstandingRequest?)
+  typealias OutboundOut = ByteBuffer
+
+  func handlerAdded(context: ChannelHandlerContext) {
+    self.scratchBuffer = context.channel.allocator.buffer(capacity: 512)
+    self.jsonDecoder.userInfo[.responseTypeCallbackKey] = { id in
+      context.eventLoop.preconditionInEventLoop()
+      guard let outstanding = self.outstandingRequests[id] else {
+        log("Unknown request for \(id)", level: .error)
+        return nil
+      }
+      return outstanding.responseType
+    } as JSONRPCMessage.ResponseTypeCallback
+  }
+
+  func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+    let bytes = unwrapInboundIn(data)
+    do {
+      let message = try self.jsonDecoder.decode(JSONRPCMessage.self, from: bytes)
+      let result: LSPResult<Any>
+      let requestID: RequestID
+
+      switch message {
+        case .errorResponse(let error, id: let id):
+          result = .failure(error)
+          requestID = id
+        case .response(let value, id: let id):
+          result = .success(value)
+          requestID = id
+        default:
+          handlerQueue.async {
+            self.messageHandler(message)
+          }
+          return
+      }
+
+      guard let outstanding = outstandingRequests.removeValue(forKey: requestID) else {
+        log("Unknown request for \(requestID)", level: .error)
+        return
+      }
+
+      outstanding.promise.succeed(result)
+
+    } catch {
+      self.errorCaught(context: context, error: error)
+    }
+  }
+
+  func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+    let (message, replyInfo) = self.unwrapOutboundIn(data)
+    if case .request(_, let id) = message {
+      precondition(outstandingRequests[id] == nil)
+      outstandingRequests[id] = replyInfo!
+    }
+    writeImpl(context: context, message: message, promise: promise)
+  }
+
+  func writeImpl(context: ChannelHandlerContext, message: JSONRPCMessage, promise: EventLoopPromise<Void>?) {
+    do {
+      self.scratchBuffer.clear()
+      try jsonEncoder.encode(message, into: &self.scratchBuffer)
+      context.write(self.wrapOutboundOut(scratchBuffer), promise: promise)
+    } catch {
+      // If anything goes wrong, tell the `Channel` and fail the write promise.
+      context.fireErrorCaught(error)
+      promise?.fail(error)
+    }
+  }
+
+  // FIXME: when can this happen?
+  func channelInactive(context: ChannelHandlerContext) {
```

Comment:
I think this FIXME came from me :-)  What does `channelInactive` mean for a pipe channel?

Reply:
@benlangmuir it means either the other side closed the input or we closed the output. If you don't enable the `ChannelOptions.allowRemoteHalfClosure` in SwiftNIO, then closing one direction automatically means both are closed. If we do enable that option, then `channelInactive` will only happen when both directions are closed individually, ie. we do receive an EOF from the other side and we do close the output.

Reply:
it's actually the same on a normal channel. Most programs don't want to think about TCP half-closure, therefore we map the closure if either direction to the closure of the whole `Channel` (unless you enable `ChannelOptions.allowRemoteHalfClosure `)

Reply:
Makes sense, thanks!  Previously when the pipe was closed we would call `close()`/`_close()` on `JSONRPCConection` , but it looks like that may have gotten lost here.

Reply:
@benlangmuir good spot! I brought that back, if the `Channel` closes, we'll now call `JSONRPCConnection._close` (on the right queue).